crates/edr_napi/src/trace/error_inferrer.rs

```diff
@@ -0,0 +1,2426 @@
+use std::{borrow::Cow, collections::HashSet};
+
+use alloy_dyn_abi::{DynSolValue, JsonAbiExt};
+use edr_evm::hex;
+use napi::{
+    bindgen_prelude::{BigInt, Either24, Either3, Either4},
+    Either, Env,
+};
+use napi_derive::napi;
+use semver::{Version, VersionReq};
+
+use super::{
+    exit::ExitCode,
+    message_trace::{CallMessageTrace, CreateMessageTrace, EvmStep, PrecompileMessageTrace},
+    model::{Bytecode, ContractFunction, ContractType, SourceLocation},
+    opcodes::Opcode,
+    return_data::ReturnData,
+    solidity_stack_trace::{
+        CallFailedErrorStackTraceEntry, CallstackEntryStackTraceEntry, CustomErrorStackTraceEntry,
+        FallbackNotPayableErrorStackTraceEntry, InternalFunctionCallStackEntry,
+        InvalidParamsErrorStackTraceEntry, NonContractAccountCalledErrorStackTraceEntry,
+        PanicErrorStackTraceEntry, SolidityStackTrace, SolidityStackTraceEntry,
+        SolidityStackTraceEntryExt, SourceReference, UnmappedSolc063RevertErrorStackTraceEntry,
+    },
+};
+use crate::{
+    trace::{
+        model::{ContractFunctionType, Instruction, JumpType},
+        solidity_stack_trace::{
+            ContractCallRunOutOfGasError, ContractTooLargeErrorStackTraceEntry,
+            DirectLibraryCallErrorStackTraceEntry,
+            FallbackNotPayableAndNoReceiveErrorStackTraceEntry,
+            FunctionNotPayableErrorStackTraceEntry, MissingFallbackOrReceiveErrorStackTraceEntry,
+            OtherExecutionErrorStackTraceEntry, ReturndataSizeErrorStackTraceEntry,
+            RevertErrorStackTraceEntry, StackTraceEntryTypeConst,
+            UnrecognizedFunctionWithoutFallbackErrorStackTraceEntry, CONSTRUCTOR_FUNCTION_NAME,
+            FALLBACK_FUNCTION_NAME, RECEIVE_FUNCTION_NAME,
+        },
+    },
+    utils::ClassInstanceRef,
+};
+
+/// Specifies whether a heuristic was applied and modified the stack trace.
+///
+/// Think of it as happy [`Result`] - the [`Heuristic::Hit`] should be
+/// propagated to the caller.
+#[must_use]
```

Comment:
While it's not idiomatic code yet, I wanted to model what was happening in the error inferrer, i.e. if a heuristic was hit, then we unconditionally propagate the modified stack trace to the caller.

However, rather than unconditionally copying the entire stack trace even when we modify it (sometimes the modifications are not big, i.e. pushing a frame), I decided to move the traces in/out of the calls and keep the result information in the discriminant; however, because we don't rely on the side-effects of those heuristics and you generally should consume the result of these calls to correctly handle this (i.e. copying the stack trace, moving it as part of the heuristic call and not handling the result is a logic error).

This is a complicated way to say that it's not strictly necessary but it makes correctly handling the pattern easier to do :sweat_smile: